dyndns/domain_update.py

```diff
@@ -27,37 +28,98 @@ def main(domain, settings='settings.txt', ignore_previous_ip=False):
+            print("Recently used {} address: {}".format(proto, ip[proto]))
+
     else:
+        answers = {}
+        for proto in protocols:
+
+            ip[proto] = None
+            record_type = protocols[proto]['record_type']
+            try:
+                answers[record_type] = dns.resolver.query(domain, record_type)
+                if not answers[record_type]:
+                    return "No {} record found for domain {}".format(record_type, domain)
+                ip[proto] = answers[record_type][0]
+                print("IP {} found in {} record".format(ip[proto], record_type))
+
+            except Exception:
+                print("No {} record found for domain {}".format(record_type, domain))
+
+        config[domain]['last_dns_check'] = int(time.time())
+
+    public_ip = {}
+    for proto in protocols:
+
+        # get public ip
+        response = requests.get(protocols[proto]['api'], params={'format': 'json'})
         try:
-            answers = dns.resolver.query(domain, 'A')
-            if not answers:
-                return "No A record found for domain {}".format(domain)
-            ip = answers[0]
-            print("IP {} found in A record".format(ip))
-            config[domain]['last_dns_check'] = int(time.time())
-        except Exception:
-            print("No A record found for domain {}".format(domain))
-
-    # get public ip
-    response = requests.get("https://api.ipify.org", params={'format': 'json'})
-    if response.status_code != 200:
-        return "Could not discover public IP."
-    public_ip = response.json().get('ip', None)
-    if not public_ip:
-        return "Could not discover public IP."
-
-    print("New IP: {}".format(public_ip))
-    if not ignore_previous_ip and ip and str(ip) == str(public_ip):
-        config[domain]['ip'] = public_ip
-        # update settings
-        with open("settings.txt", "w") as settings_file:
-            json.dump(config, settings_file, sort_keys=True, indent=1)
-        return "A record up to date."
+            response_ip = response.json().get('ip', None)
+            # validate http response code
+            if response.status_code != 200:
+                raise ValueError("Could not discover public {} address.".format(proto))
+
+            # validate the correct returned IP Version
+            if ipaddress.ip_address(response_ip.decode('unicode-escape')).version != protocols[proto]['version']:
+                raise ValueError('The recieved IP address {} doesn\'t match the Protocol {}'
+                                 .format(response_ip, proto))
+
+            # validation ok
+            public_ip[proto] = response_ip
+            print("New {} address: {}".format(proto, public_ip[proto]))
+
+        except ValueError as error:
+            print(error)
+            # validation failed! do we have a valid ip elsewhere?
+            # if whe got an ip from the dns query, use it
+            if proto in ip and ip[proto]:
```

Comment:
Not sure the intention of this fallback. `ip[proto]` holds a DNS response for a record type.
If it's assigned to `public_ip[proto]` it may have strange side-effects, like not recognized a needed update (public_ip==ip) or updating an old value (ip from DNS not yet propagated and responding still with old value due to caching).

Reply:
I made these fallbacks, because if you have set up both ipv4 and ipv6, you have to provide both in the update request. Sometimes it coud be, that no valid ipv6 address could be fetched, so to update at least the new ipv4, you still have to give a valid ipv6 address. 

Following case: We set up ipv4 AND ipv6 and want to update now. We retrieve a new ipv4 from ipify, but temporarely can't get a new ipv6 address. So its still better to update only ipv4 by providing an old ipv6, than to abort and do not update anything.

But I can understand possible side effects. Do you propose to get rid of the two fallbacks and fail?

Reply:
OK, in this case fallback to DNS value is not a desired solution, as in the worst case it may mean updating to an old value if public_ip won't resolve (DNS may still hold olf value due to cache).
The fallback to last update value may be a valid approach.